REGRESSION(295435@main): WPE Platform: crash under Wayland with software rasterization

carlosgcampos opened the pull request:
#### f528ed283464aa5312099450789410c81b9b04a7
<pre>
REGRESSION(295435@main): WPE Platform: crash under Wayland with software rasterization
<a href="https://bugs.webkit.org/show_bug.cgi?id=294718">https://bugs.webkit.org/show_bug.cgi?id=294718</a>

Reviewed by Nikolas Zimmermann.

In 295435@main we changed the way we release the WPEBuffer on wayland
buffer listener assuming the WPEBuffer is always a DMABuf. We need to do
the same changes for shared memory buffers.

* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(SharedMemoryBuffer::SharedMemoryBuffer):
(SharedMemoryBuffer::~SharedMemoryBuffer):
(sharedMemoryBufferDestroy):
(sharedMemoryBufferCreate):
(createWaylandBufferSHM):
(wpeViewWaylandSetCursorFromBytes):

Canonical link: <a href="https://commits.webkit.org/296416@main">https://commits.webkit.org/296416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1194a59e7f60219912c7296233611e3398ce0f5b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108458 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18543 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113667 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58861 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28808 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36669 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82362 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111406 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/22849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97689 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62798 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/22265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58387 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15880 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116788 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35512 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91387 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35885 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93963 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/91188 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36081 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13843 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31258 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17516 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35414 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40949 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35124 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38475 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36805 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -158,13 +158,44 @@ static void dmaBufBufferDestroy(DMABufBuffer* buffer)
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
+
+    SharedMemoryBuffer(WPEView* view, std::unique_ptr<WPE::WaylandSHMPool>&& pool, uint32_t offset, uint32_t width, uint32_t height, uint32_t stride)
+        : view(view)
```
wpeView(view) similar to wlPooo(pool)?

carlosgcampos replied:
Here we are consistent with DMABufBuffer. I plan to use a base class for both so that we don't need to check the buffer type to get its view, for example.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -158,13 +158,44 @@ static void dmaBufBufferDestroy(DMABufBuffer* buffer)
+
+    GWeakPtr<WPEView> view;
+    std::unique_ptr<WPE::WaylandSHMPool> wlPool;
+    struct wl_buffer* wlBuffer;
```
{ nullptr } for consistency

carlosgcampos committed:
REGRESSION(295435@main): WPE Platform: crash under Wayland with software rasterization
https://bugs.webkit.org/show_bug.cgi?id=294718

Reviewed by Nikolas Zimmermann.

In 295435@main we changed the way we release the WPEBuffer on wayland
buffer listener assuming the WPEBuffer is always a DMABuf. We need to do
the same changes for shared memory buffers.

* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(SharedMemoryBuffer::SharedMemoryBuffer):
(SharedMemoryBuffer::~SharedMemoryBuffer):
(sharedMemoryBufferDestroy):
(sharedMemoryBufferCreate):
(createWaylandBufferSHM):
(wpeViewWaylandSetCursorFromBytes):

Canonical link: https://commits.webkit.org/296416@main
